### PR TITLE
Interface

### PR DESCRIPTION
**Pull request by yumuzi234**

### CODE REVIEW

File: pl/sempass/assign.go [hunk 1]
```diff
@@ -10,7 +10,7 @@ import (
 func assign(b *builder, dest, src tast.Expr, op *lexing.Token) tast.Stmt {
 	destRef := dest.R()
 	srcRef := src.R()
-
+	isError := false
```
Comment (h8liu) [inline pl/sempass/assign.go]: rename to `seenError`

File: pl/sempass/assign.go [hunk 2]
```diff
@@ -30,13 +30,16 @@ func assign(b *builder, dest, src tast.Expr, op *lexing.Token) tast.Stmt {
```
Comment (h8liu) [inline pl/sempass/assign.go]: also replace the previous `return nil`

File: pl/sempass/assign.go [hunk 3]
```diff
@@ -139,3 +142,90 @@ func buildAssignStmt(b *builder, stmt *ast.AssignStmt) tast.Stmt {
 
 	return opAssign(b, left, right, stmt.Assign)
 }
+
+func canAssign(b *builder, p *lexing.Pos,
+	left types.T, right types.T) bool {
```
Comment (h8liu) [inline pl/sempass/assign.go]: `left, right types.T`; skip the `types.T` in the middle.

File: pl/sempass/assign.go [hunk 4]
```diff
@@ -139,3 +142,90 @@ func buildAssignStmt(b *builder, stmt *ast.AssignStmt) tast.Stmt {
...
+	if i, ok := left.(*types.Interface); ok {
+		// TODO
```
Comment (h8liu) [inline pl/sempass/assign.go]: `TODO(yumuzi234): something description`

File: pl/sempass/assign.go [hunk 5]
```diff
@@ -139,3 +142,90 @@ func buildAssignStmt(b *builder, stmt *ast.AssignStmt) tast.Stmt {
...
+	flag := true
+	s, ok := types.PointerOf(right).(*types.Struct)
+	if !ok {
+		b.CodeErrorf(p, "pl.cannotAssign.interface",
+			"cannot assign interface %s by %s, not a struct pointer", i, right)
+		return false
+	}
+	e := func(fname, m string) {
+		b.CodeErrorf(p, "pl.cannotAssign.interface",
+			"cannot assign interface %s by %s, "+m, i, right, fname)
```
Comment (h8liu) [inline pl/sempass/assign.go]: rename `e` to something like `errorf`.
rename m to `format` and put it as the first argument, change the second one into `arg... interface{}`

Comment (h8liu) [inline pl/sempass/assign.go]: then, inside this function, use `fmt.Sprintf` to generate the later part first, and then generate the full error using `CodeErrorf`.

File: pl/sempass/call.go [hunk 6]
```diff
@@ -144,27 +148,32 @@ func buildCallExpr(b *builder, expr *ast.CallExpr) tast.Expr {
 
 	argsRef := args.R()
 	nargs := argsRef.Len()
+	p := ast.ExprPos(expr)
```
Comment (h8liu) [inline pl/sempass/call.go]: let's use `pos`

File: pl/sempass/call.go [hunk 7]
```diff
@@ -144,27 +148,32 @@ func buildCallExpr(b *builder, expr *ast.CallExpr) tast.Expr {
...
 	if nargs != len(funcType.Args) {
-		b.CodeErrorf(ast.ExprPos(expr), "pl.argsMismatch.count",
+		b.CodeErrorf(p, "pl.argsMismatch.count",
 			"argument count mismatch, expects (%s), got (%s)",
 			fmtutil.Join(funcType.Args, ","), args,
 		)
 		return nil
 	}
-
+	isError := false
```
Comment (h8liu) [inline pl/sempass/call.go]: `seenError`

File: pl/sempass/member.go [hunk 8]
```diff
@@ -116,31 +117,38 @@ func buildMember(b *builder, m *ast.MemberExpr) tast.Expr {
...
-	pt := types.PointerOf(t)
-	var tstruct *types.Struct
-	var ok bool
-	if pt != nil {
-		if tstruct, ok = pt.(*types.Struct); !ok {
-			b.Errorf(m.Dot.Pos, "*%s is not a pointer of struct", t)
-			return nil
-		}
+	var symTable *syms.Table
+	// TODO: Interface
```
Comment (h8liu) [inline pl/sempass/member.go]: say more about the todo, try use a full sentence. also, add your name:
`TODO(yumuzi234):`

File: pl/sempass/assign.go [hunk 9]
```diff
@@ -56,6 +63,20 @@ func assign(b *builder, dest, src tast.Expr, op *lexing.Token) tast.Stmt {
 		src = newList
 	}
 
+	if destList, ok := tast.MakeExprList(dest); ok {
+		newList := tast.NewExprList()
+		for i, e := range destList.Exprs {
+			if v, ok := e.Type().(*types.Interface); ok {
+				e = interfaceCast(b, nil, srcList.Exprs[i], v)
```
Comment (yumuzi234) [inline pl/sempass/assign.go]: for this tast.Cast, it needs the expr for the struct, and also the expr of the interface, only type is note enough. I think.
May I use Ref? probably not a good idea.

File: pl/sempass/cast.go [hunk 10]
```diff
@@ -55,3 +57,8 @@ func buildCast(b *builder, expr *ast.CallExpr, t types.T) tast.Expr {
 	b.Errorf(pos, "cannot convert from %s to %s", srcType, t)
 	return nil
 }
+
+func interfaceCast(b *builder, pos *lexing.Pos, from tast.Expr,
```
Comment (yumuzi234) [inline pl/sempass/cast.go]: Here what we need to change it the syms of the interface, right?

File: pl/sempass/assign.go (hunk 9 above)
Comment (h8liu) [inline pl/sempass/assign.go]: you should use ref for the struct. can you explain more on why you need the Ref of the interface?

File: pl/sempass/cast.go (hunk 10 above)
Comment (h8liu) [inline pl/sempass/cast.go]: could you explain more? what do you mean by "it" in "change it the syms of the interface"?

File: pl/cast.go [hunk 11]
```diff
@@ -49,8 +49,11 @@ func buildCast(b *builder, from *ref, t types.T) *ref {
 		if size == arch.RegSize {
 			return newRef(t, codegen.Num(0))
 		}
+		if _, ok := t.(*types.Interface); ok {
+			panic("Interface is not supported yet")
+		}
 		if _, ok := t.(*types.Slice); !ok {
-			panic("bug")
+			panic("unknow type")
```
Comment (h8liu) [inline pl/cast.go]: unknown

File: pl/cast.go [hunk 12]
```diff
@@ -49,8 +49,11 @@ func buildCast(b *builder, from *ref, t types.T) *ref {
 		if size == arch.RegSize {
 			return newRef(t, codegen.Num(0))
 		}
+		if _, ok := t.(*types.Interface); ok {
+			panic("Interface is not supported yet")
```
Comment (h8liu) [inline pl/cast.go]: so here is a basic rule: don't panic the compiler. only use panic when the code is impossible to reach. `pl.Builder` also has an error list; please use that to generate an error as a place holder here.

one reason for this rule is that when the compiler is compiled into js and runs in a browser, panic will kill the gopherjs runtime.

File: pl/sempass/assign.go [hunk 13]
```diff
@@ -139,3 +138,90 @@ func buildAssignStmt(b *builder, stmt *ast.AssignStmt) tast.Stmt {
 
 	return opAssign(b, left, right, stmt.Assign)
 }
+
+func canAssign(b *builder, p *lexing.Pos, left, right types.T) bool {
+	if i, ok := left.(*types.Interface); ok {
+		// TODO(yumuzi234): assing interface from interface
```
Comment (h8liu) [inline pl/sempass/assign.go]: what is "assing"?

File: pl/sempass/assign.go [hunk 14]
```diff
@@ -139,3 +138,90 @@ func buildAssignStmt(b *builder, stmt *ast.AssignStmt) tast.Stmt {
...
+func assignInterface(b *builder, p *lexing.Pos,
+	i *types.Interface, right types.T) bool {
+	flag := true
+	s, ok := types.PointerOf(right).(*types.Struct)
+	if !ok {
+		b.CodeErrorf(p, "pl.cannotAssign.interface",
+			"cannot assign interface %s by %s, not a struct pointer", i, right)
+		return false
+	}
+	errorf := func(format string, arg ...interface{}) {
```
Comment (h8liu) [inline pl/sempass/assign.go]: replace `arg` with `args`. it is a slice. or just use `f` and `a`

File: pl/sempass/cast.go [hunk 15]
```diff
@@ -51,7 +53,29 @@ func buildCast(b *builder, expr *ast.CallExpr, t types.T) tast.Expr {
...
 	}
+	if _, ok := t.(*types.Interface); ok {
+		return tast.NewCast(args, t)
+	}
 
 	b.Errorf(pos, "cannot convert from %s to %s", srcType, t)
 	return nil
 }
+
+func implicitTypeCast(
```
Comment (h8liu) [inline pl/sempass/cast.go]: let's call this `implicitCast`.

**Pull request merged**